app/javascript/lib/validation/validate-file-content.js

```diff
@@ -266,6 +281,104 @@ function validateNoMetadataCoordinates(headers) {
   return issues
 }
 
+/** Verifies metadata file has all required columns */
+function validateRequiredMetadataColumns(parsedHeaders) {
+  const issues = []
+  const firstLine = parsedHeaders[0]
+  const missingCols = []
+  REQUIRED_CONVENTION_COLUMNS.forEach(colName => {
+    if (!firstLine.includes(colName)) {
+      missingCols.push(colName)
+    }
+  })
+  if (missingCols.length) {
+    const msg = `File is missing required columns ${missingCols.join(', ')}`
+    issues.push(['error', 'format:cap:metadata-missing-column', msg])
+  }
+  return issues
+}
+
+/**
+ * Verify that, for id columns with a corresponding label column, no label is shared across two or more ids.
+ * The main circumstance this is aimed at checking is the 'Excel drag error', in which by drag-copying a row, the
+ * label is copied correctly, but the id string gets numerically incremented
+ */
+function validateMetadataLabelMatches(headers, line, isLastLine, dataObj) {
+  const issues = []
+  const excludedColumns = ['NAME']
+  // if this is the first time through, identify the colums to check, and initialize data structures to track mismatches
+  if (!dataObj.dragCheckColumns) {
+    dataObj.dragCheckColumns = headers[0].map((colName, index) => {
+      const labelColumnIndex = headers[0].indexOf(`${colName }__ontology_label`)
+      if (excludedColumns.includes(colName) ||
+        colName.endsWith('ontology_label') ||
+        headers[1][index] === 'numeric' ||
+        labelColumnIndex === -1) {
+        return null
+      }
+      // for each column, track a hash of label=>value,
+      // and also a set of mismatched values--where the same label is used for different ids
+      return { colName, index, labelColumnIndex, labelValueMap: {}, mismatchedVals: new Set() }
+    }).filter(c => c)
+  }
+  // for each column we need to check, see if there is a corresponding filled-in label,
+  //  and track whether other ids have been assigned to that label too
+  dataObj.dragCheckColumns.forEach(dcc => {
+    const colValue = line[dcc.index]
+    const label = line[dcc.labelColumnIndex]
+    if (label.length) {
+      if (dcc.labelValueMap[label] && dcc.labelValueMap[label] !== colValue) {
+        dcc.mismatchedVals.add(label)
+      } else {
+        dcc.labelValueMap[label] = colValue
+      }
+    }
+  })
```

Comment:
Blocker: please use `for` instead of `forEach`, as this is called many times.  It'd [iterate ~2.5x faster](https://leanylabs.com/blog/js-forEach-map-reduce-vs-for-for_of/#arrayforeach-vs-for-and-forof).